### DISCUSSION TITLE
What causes the observer to re-render a component?

### DISCUSSION THREAD

**Discussion by joaovct** [mobxjs/mobx#3244] [Q&A] [answered]
Have been playing around a little with the `observer` and found out to be interesting the way it avoids unnecessary re-renders. However, I want to know more about it. The [docs](https://mobx.js.org/react-integration.html#react-integration) says: "It also makes sure that components don't re-render when there are _no relevant_ changes?". My question it's: what does "no relevant changes" really means?

**Comment by urugator** [comment, marked as answer]
Generally speaking if it wouldn't change the output it's not relevant, eg:
```javascript
// Conditions
observer(({ store }) => {  
  return store.cond ? store.a : store.b
  // if `store.cond` is `true`, `store.b++` is NOT relevant
})

// Object props
observer(({ store }) => {
  return store.a;
  // `store.b++` is NOT relevant
  // `store.a++` IS relevant
})

// Passthrough
observer(({ store }) => {    
  return <User user={store.users[0]}/>
  // `store.users[0].name = "new name"` is NOT relevant for this component
  // it can be relevant for User component, so it may need to be observer
  // `store.users = [{ name: "foo" }]` IS relevant
  // `store.users[0] = { name: "foo" }` IS relevant
})
```
Note in all above cases, swapping `store` itself is relevant, but that's simply handled by `React.memo`.

The logic is simple: if value was accessed, it will affect the result, therefore is relevant (assumes there are no unused vars or meaningless expressions).
It's implemented by injecting subscription/notification logic into getters/setters or via proxy handlers +  global state.

https://mobx.js.org/understanding-reactivity.html

**Reply by joaovct** [reply]
So, if we had a component like this:

````.ts
const List = observer(({ list }) => {
  // don't have any other internal state
  console.log(list);

  return ...
  );
});
````

Can we correctly affirm that the console.log will only be prompted when `list` updates, besides whatever happens in the components above it?

**Reply by urugator** [reply]
Why don't you try?

> whatever happens in the components above it

`observer` implicitly applies `memo`, so if `list` is the same reference (which is usually the case with mutable state), then shallow equality check yields true, skipping render...

**Reply by joaovct** [reply]
Try it before and notice this behavior. Just wanted to be 100% sure.
